### PR TITLE
[openwebnet] add support for Energy Meter

### PR DESCRIPTION
**Pull request by aconte80**
These changes will:

- restore support for Energy Meters (formerly "Energy Central Unit" in OH2, for example BTI-F520). [Issue #10188 ](https://github.com/openhab/openhab-addons/issues/10188) or see [community thread](https://community.openhab.org/t/bticino-openwebnet-new-openhab2-binding-ready-for-testing/43377/1444).
- Improvement: handle readed data with Unit of Measure (Watt)

Signed-off-by: Andrea Conte <andrea@conte.com>

### CODE REVIEW

**Comment by mvalla** [comment]
this PR will also close #10242 by using version 0.4.0 of the openwebnet4j lib (see: https://github.com/mvalla/openwebnet4j/releases/tag/0.4.0).

File: bundles/org.openhab.binding.openwebnet/README.md
Comment (fwolter) [inline bundles/org.openhab.binding.openwebnet/README.md:46]: There should be an empty line above headings. Same for below.

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/OpenWebNetBindingConstants.java
Comment (fwolter) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/OpenWebNetBindingConstants.java:82]: Same for below.
```suggestion
    // Energy Management
```

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetBridgeHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetBridgeHandler.java:295]: Can you fix the formatting of the comment?

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java:65]: This is logged by the framework and can therefore be removed. Same for below.

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java:82]: Bindings should only log to error if something severe happened, like the detection of a bug in your code. This could be debug. See this link for a description of the log levels: https://www.openhab.org/docs/developer/guidelines.html#f-logging

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java:98]: Can this message be replaced by using the debugger or by increasing the framework's log level? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java:118]: Can this message be replaced by using the debugger or by increasing the framework's log level? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java:116]: Can this be removed?

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java:128]:
```suggestion
}
```

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/BusEnergyMeter.xml
Comment (fwolter) [inline bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/BusEnergyMeter.xml:30]: Labels are expected to be as short as possible. Guideline is 2-3 words with max 20-25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml
Comment (fwolter) [inline bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:42]:
```suggestion
		<item-type>Number:Power</item-type>
```

**Review by fwolter — CHANGES_REQUESTED** [review]
Did you modify the images intentionally?

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/OpenWebNetBindingConstants.java
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/OpenWebNetBindingConstants.java:82]: All comments in other sections begin with ##.
line 71: // ## Generic
line 74: // ## Lighting
line 78: // ## Automation
...and so on.
Should I leave it as is or do you want me to modify them all?

Comment (fwolter) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/OpenWebNetBindingConstants.java:82]: Seems a bit strange with the double pound. I leave it up to you if you change all or keep all.

Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/OpenWebNetBindingConstants.java:82]: I prefere to keep them all.

File: bundles/org.openhab.binding.openwebnet/README.md
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/README.md]: added empty line

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetBridgeHandler.java
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetBridgeHandler.java]: removed extra line returns

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java]: changed to warn level. This should never happens but in case the flow continues.

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java]: removed

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java]: removed, it was a typo

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java]: removed as suggested

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java]: removed

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java]: removed

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/BusEnergyMeter.xml
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/BusEnergyMeter.xml]: changed to "OpenWebNet Address"

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml]: fixed

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java:71]: What's the purpose of these?

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetEnergyHandler.java]: It's a mistake, initialize() and dispose() aren't declared abstract so it's not mandatory to override them.

**Comment by fwolter** [comment]
@mvalla The 0.4.0 version is not found when building this PR.

**Comment by mvalla** [comment]
> @mvalla The 0.4.0 version is not found when building this PR.

@fwolter Very strange: actually all versions of the lib are now returning "403 Fobidden".
For example: https://jcenter.bintray.com/com/github/openwebnet4j/openwebnet4j/0.3.4/

Maybe it's a temporary problem, hopefully not related to the soon to be [Bintray/JCenter shutdown](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/)

**Comment by kaikreuzer** [comment]
> Maybe it's a temporary problem, hopefully not related to the soon to be Bintray/JCenter shutdown

It is related to it. I have started working on https://github.com/openhab/openhab-distro/issues/1256 and this requires us to not use any dependencies from JCenter anymore.

@mvalla Could you publish the lib to Maven Central instead? This seems to be the only working option on the long run.
As long as Bintray is still available, you can use Bintray's "Push to Maven Central" feature to do it without too much effort.

**Comment by mvalla** [comment]
it looks like a double problem. 
1. artifacts are not accessible anymore: they all return 403 Forbidden. After a quick check this applies to [other artifacts as well - just click on one of the files](https://bintray.com/bintray/jcenter/com.discoverydns.dnsapi%3Adnsapi-client/1.4.0#files/com/discoverydns/dnsapi/dnsapi-client). I have seen some [service shortage announced on JCenter](https://status.bintray.com/incidents/9n84kbm0bjh0), but now declared as solved. This is strange as [they promised](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/) existing artifacts will be served until 2022:

> After this date JCenter repository will still resolve artifacts, but end users should migrate to the canonical repo (e.g. Java packages should be retrieved from Maven Central). On Feb 1st, 2022 JCenter will only resolve artifacts for Artifactory clients.

2. migration away from JCenter by @kaikreuzer : yes of course I would like to move the lib to Maven Central, but it's not a 1-click process and the quicker "Push to Maven Central" does not work atm, maybe because of the 403 Forbidden error. I will have to go through the Sonatype approval process, and if everybody is migrating now they will be surely very busy (and slow). 

Did you have similar problems with other artifacts used by OH and not available anymore?

**Comment by kaikreuzer** [comment]
Yes, I have similar problems with other artifacts and I am looking for an alternative source for each. If none can be found, I temporarily upload them to our Artifactory. I just did so with openwebnet4j as well, but I uses `openwebnet4j` as its group id.
Only noticed now that you defined the group id to be `com.github.openwebnet4j`. Why this, this is not at Github library, is it? I strongly suggest to change the groupId because otherwise you'll anyhow not be allowed to bring it to Maven Central.

File: bundles/org.openhab.binding.openwebnet/pom.xml
Comment (kaikreuzer) [inline bundles/org.openhab.binding.openwebnet/pom.xml:24]:
```suggestion
      <groupId>openwebnet4j</groupId>
```
This should make it resolve from our Artifactory for now.

**Comment by mvalla** [comment]
Thanks @kaikreuzer for hosting the artifact on OH Artifactory : it's a shame JCenter apparently is offline even before the planned date!

> Only noticed now that you defined the group id to be `com.github.openwebnet4j`. Why this, this is not at Github library, is it? I strongly suggest to change the groupId because otherwise you'll anyhow not be allowed to bring it to Maven Central.

Actually [the policy is](https://central.sonatype.org/pages/producers.html#individual-projects-open-source-software-repository-hosting-ossrh):
> you must choose a groupId for a domain that you own, or for which you are the designated maintainer on behalf of the owner. In the case of a GitHub groupId (io.github.username), this will be immediately verified provided your project URL matches the requested groupId.

that is why I used `com.github.openwebnet4j` in the first place and I also created the github group `openwebnet4j` to migrate in the future.

So to summarize: I will move to Maven Central keeping `com.github.openwebnet4j` but it takes time...

**Comment by kaikreuzer** [comment]
> that is why I used `com.github.openwebnet4j` in the first place

If you read carefully, it says that the groupId must be `io.github.mvalla`. Maven Central [seems to be fine](https://issues.sonatype.org/browse/OSSRH-64495) with using `com` instead of `io` as well, but as this is not specified in their guideline, I'd go for `io`, since this is what is used for Github pages (and hence it qualifies as a domain name that is under your control).

**Comment by kaikreuzer** [comment]
@mvalla I have uploaded version 0.4.0 to our Artifactory under the current groupId, so that it can now be resolved.
Please adapt the groupId for future releases, though.
The build now fails with spotless errors that you'll need to fix as it seems.

**Comment by mvalla** [comment]
> @mvalla I have uploaded version 0.4.0 to our Artifactory under the current groupId, so that it can now be resolved.
> Please adapt the groupId for future releases, though.
> The build now fails with spotless errors that you'll need to fix as it seems.

thanks @kaikreuzer for publishing under OH Artifactory with the current groupId: that is indeed the cleanest temporary solution.
JCenter is still returning 403 Forbidden (for past releases as well).
I will open a new ticket to publish via OSSRH to Maven Central with new groupId `io.github.openwebnet4j`.

I leave to @aconte80 the fixes for the spotless errors.

**Comment by kaikreuzer** [comment]
> I leave to @aconte80 the fixes for the spotless errors.

Ah, right, I didn't notice that this PR wasn't from you :-)

**Comment by aconte80** [comment]
@kaikreuzer @fwolter I've fixed the spotless error.

**Review by kaikreuzer — APPROVED** [review]
Perfect, thank you all!

**Comment by mvalla** [comment]
@fwolter @wborn please add link to issue #10242 that is closed by this PR

**Comment by wborn** [comment]
They are linked now @mvalla. :-)

**Pull request merged**
